server/app/filters/SessionTimeoutFilter.java

```diff
@@ -0,0 +1,141 @@
+            maybeSessionStartTime -> {
+              // Get timestamps in seconds
+              long currentTime = clock.instant().toEpochMilli() / 1000;
+              long sessionStartTimeInMillis = maybeSessionStartTime.orElse(currentTime);
+              long lastActivityTime = profile.getProfileData().getLastActivityTime(clock) / 1000;
+              long sessionStartTime = sessionStartTimeInMillis / 1000;
+
+              // Get configuration values
+              int inactivityMinutes =
+                  settingsManifest
+                      .get()
+                      .getSessionInactivityTimeoutMinutes()
+                      .orElse(DEFAULT_INACTIVITY_TIMEOUT_MINUTES);
+              int totalLengthMinutes =
+                  settingsManifest
+                      .get()
+                      .getMaximumSessionDurationMinutes()
+                      .orElse(DEFAULT_MAX_SESSION_DURATION_MINUTES);
+              int inactivityWarningMinutes =
+                  settingsManifest.get().getSessionInactivityWarningThresholdMinutes().orElse(5);
+              int durationWarningMinutes =
+                  settingsManifest.get().getSessionDurationWarningThresholdMinutes().orElse(10);
+
+              // Calculate limits
+              long inactivityLimit =
+                  Instant.ofEpochSecond(lastActivityTime)
+                      .plus(inactivityMinutes, ChronoUnit.MINUTES)
+                      .getEpochSecond();
+              long totalLengthLimit =
+                  Instant.ofEpochSecond(sessionStartTime)
+                      .plus(totalLengthMinutes, ChronoUnit.MINUTES)
+                      .getEpochSecond();
+
+              // Calculate warnings
+              long inactivityWarning =
+                  Instant.ofEpochSecond(inactivityLimit)
+                      .minus(inactivityWarningMinutes, ChronoUnit.MINUTES)
+                      .getEpochSecond();
+              long totalLengthWarning =
+                  Instant.ofEpochSecond(totalLengthLimit)
+                      .minus(durationWarningMinutes, ChronoUnit.MINUTES)
+                      .getEpochSecond();
+
+              ObjectNode timestamps =
+                  Json.newObject()
+                      .put("inactivityWarning", inactivityWarning)
+                      .put("inactivityTimeout", inactivityLimit)
+                      .put("totalWarning", totalLengthWarning)
+                      .put("totalTimeout", totalLengthLimit)
+                      .put("currentTime", currentTime);
+
+              String cookieValue =
+                  Base64.getEncoder()
+                      .encodeToString(Json.stringify(timestamps).getBytes(StandardCharsets.UTF_8));
+
+              // Set Cookie with appropriate max age
+              return Http.Cookie.builder(TIMEOUT_COOKIE_NAME, cookieValue)
+                  .withHttpOnly(false)
+                  .withPath("/")
+                  .withMaxAge(COOKIE_MAX_AGE)
+                  .build();
```

Comment:
(question) Will you be using javascript to read the values of the cookie on the client side?

Reply:
yes. That is correct. See session.ts here: https://github.com/civiform/civiform/pull/9776